server/src/main/java/org/opensearch/action/search/QueryInsightService.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.lifecycle.AbstractLifecycleComponent;
+import org.opensearch.threadpool.Scheduler;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting data related to
+ * search queries, based on certain dimensions.
+ *
+ * @param <R> The type of record that stores in the service
+ * @param <S> The type of Collection that holds the aggregated data
+ * @param <E> The type of exporter that exports the aggregated and processed data
+ *
+ * @opensearch.internal
+ */
+public abstract class QueryInsightService<R extends SearchQueryRecord<?>, S extends Collection<R>, E extends QueryInsightExporter<R>>
+    extends AbstractLifecycleComponent {
+    private static final Logger log = LogManager.getLogger(QueryInsightService.class);
+    private boolean enabled;
+
+    /** The internal store that holds the query insight data */
+    @Nullable
+    protected S store;
```

Comment:
can there be multiple stores?

Reply:
I think we can define one for a common query insight service? If the child services needs multiple stores they can define their private stores since it would be hard to abstract different types of stores in the abstract class.